mono/mini/exceptions-arm64.c

```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+}
+
+static void
+mono_arch_unwind_add_save_regp_x (guint8* unwind_codes, guint32* unwind_code_size, guint32 reg, guint32 offset, gboolean reverse) {
+
+	mono_arch_unwind_add_assert (2, 256, *unwind_code_size, offset);
+
+	guint8 reg_offset = reg - ARMREG_R19;
+	int order = reverse ? 1 : 0;
+
+	unwind_codes[*unwind_code_size + order] = 0b11001100 | reg_offset / 8;
+	unwind_codes[*unwind_code_size + 1 - order] = (reg_offset & 0x8) << 5 | (offset / 8);
+	*(unwind_code_size) += 2;
+}
+
+static void
+mono_arch_unwind_add_alloc_s (guint8* unwind_codes, guint32* unwind_code_size, guint32 size) {
+
+	mono_arch_unwind_add_assert (1, 511, *unwind_code_size, size);
+	unwind_codes[(*unwind_code_size)++] = 0b00000000 | (size / MONO_ARCH_FRAME_ALIGNMENT);
+}
+
+static void
+mono_arch_unwind_add_alloc_m (guint8* unwind_codes, guint32* unwind_code_size, guint32 size, gboolean reverse) {
+
+	mono_arch_unwind_add_assert (2, 0x7FFF, *unwind_code_size, size);
+	int order = reverse ? 1 : 0;
+	size = size / 16; // Size is multiples of 16
+	unwind_codes[*unwind_code_size + order] = 0b11000000 | (size >> 8);
+	unwind_codes[*unwind_code_size + 1 - order] = size & 0xFF;
+	*(unwind_code_size) += 2;
+}
+
+static void
+mono_arch_unwind_add_alloc_l (guint8* unwind_codes, guint32* unwind_code_size, guint32 size, gboolean reverse) {
+
+	mono_arch_unwind_add_assert (4, 0x0FFFFFFF, *unwind_code_size, size);
+	int order = reverse ? 1 : 0;
+
+	size = size / 16; // Size is multiples of 16
+	unwind_codes[*unwind_code_size + order * 3] = 0b11100000;
+	unwind_codes[*unwind_code_size + 1 + order] = size >> 16;
+	unwind_codes[*unwind_code_size + 2 - order] = (size >> 8) & 0xFF;
+	unwind_codes[*unwind_code_size + 3 - order * 3] = size & 0xFF;
+
+	(*unwind_code_size) += 4;
+}
+
+static void
+mono_arch_unwind_add_alloc_x (guint8* unwind_codes, guint32* unwind_code_size, guint32 size, gboolean reverse) {
+
+	if (size < 512)
+		mono_arch_unwind_add_alloc_s (unwind_codes, unwind_code_size, size);
+	else if (size < 0x8000)
+		mono_arch_unwind_add_alloc_m (unwind_codes, unwind_code_size, size, reverse);
+	else
+		mono_arch_unwind_add_alloc_l (unwind_codes, unwind_code_size, size, reverse);
+}
+
+static int
+mono_arch_unwind_add_reg_offset (MonoUnwindOp* unwind_op_data, GSList* next, gint stack_offset, guint8 *unwind_codes, guint32 *unwind_code_size, gboolean reverse) {
+
+	if (unwind_op_data->reg == ARMREG_SP) {
+		g_assert("I don't believe there is a way to encode this in the unwind info");
+		return 0;
+	}
+
+	if (unwind_op_data->reg == ARMREG_FP || unwind_op_data->reg == ARMREG_LR) {
+		// FP and LR are handled seperately
+		return 0;
+	}
+
+	if (unwind_op_data->reg >= ARMREG_R19 && unwind_op_data->reg <= ARMREG_R28) {
+		// Only the presreved registers have unwind codes
+		// We shoulded hit here, but emit a nop if we do
+		g_assert(unwind_op_data->reg >= ARMREG_R19 && unwind_op_data->reg <= ARMREG_R28);
+		mono_arch_unwind_add_nop(unwind_codes, unwind_code_size);
+		return 1;
```

Comment:
Should we have `mono_arch_unwind_add_*` methods return the size used?  

This becomes `return mono_arch_unwind_add_nop(unwind_codes, unwind_code_size);`.

Just feels safer to me

Reply:
I think I misunderstood the return value. That is the reg count